db/schema_tables.cc

```diff
@@ -947,11 +948,23 @@ static future<> with_merge_lock(noncopyable_function<future<> ()> func) {
     }
 }
 
+template <typename Func>
+requires std::same_as<futurize_t<std::invoke_result_t<Func, replica::database&>>, future<>>
+static future<> update_database(sharded<replica::database>& db, Func func) {
+    using futurator = futurize<std::invoke_result_t<Func, replica::database&>>;
+    co_await futurator::invoke(func, db.local());
+    try {
+        co_await db.invoke_on_others(func);
+    } catch (...) {
+        on_fatal_internal_error(slogger, fmt::format("Updating database schema on all shards failed: {}", std::current_exception()));
+    }
+}
```

Comment:
This patch changes the behavior in a significant way. We will abort the node on failure instead of failing the schema merge. The title should communicate that.

Reply:
The asymmetry between the way we handle shard 0 failure and non-shard-0 failure (throw vs abort) is not helpful in the more general scope. Even though this update throws with no side effects, prior calls to update_database() which completed already had side effects.

Reply:
The decision how to handle that could be a separate PR. If erm is computed only on shard 0, we don't need this patch to ensure consistency.

Reply:
It's helpful for cross-shard consistency.
I.e. if the schema change was partially applied, at least all shards should be consistent with each other.
But indeed not consistent with the latest schema.
If it's an invalid state to be in, maybe we need to consider aborting after such an error, and let the schema be applied after restart,

Reply:
I agree.  I think it would be better to abort if we can't ensure consistency, rather than deal with broken state.  This way the database objects will be repopulated after restart.

Reply:
See `database: update_keyspace_on_all_shards: abort on error` in latest version